### PR TITLE
pb: Initialise pb cli

### PR DESCRIPTION
**Pull request by juliaogris**
Create basic `pb` command converting proto message from one encoding
format to another, currently supporting JSON, prototext, binary proto.

Add basic docs, started by @camh- and goreleaser config.

### CODE REVIEW

**Review by camh- — APPROVED** [review]
:leaves: All very good - just nits and and thoughts.

File: Makefile [hunk 1]
```diff
@@ -13,18 +13,19 @@ all: build test check-coverage lint  ## build, test, check coverage and lint
 ci: clean check-uptodate all  ## Full clean build and up-to-date checks as run on CI
 
 clean::  ## Remove generated files
-	-rm -rf $(O)
+	-rm -rf $(O) base-message.pb
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply Makefile:16]: yesyes totally agreed. that was an accidental leftover. sorry!

File: Makefile [hunk 2]
```diff
@@ -13,18 +13,19 @@ all: build test check-coverage lint  ## build, test, check coverage and lint
...
 
 .PHONY: all ci clean
 
 # --- Build --------------------------------------------------------------------
 GO_LDFLAGS = -X main.version=$(VERSION)
+CMDS = ./cmd/pb
 
 build: | $(O)  ## Build reflect binaries
-	go build -o $(O)/protog -ldflags='$(GO_LDFLAGS)' .
+	go build -o $(O) -ldflags='$(GO_LDFLAGS)'  $(CMDS)
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply Makefile:25]: thanks done.

File: Makefile [hunk 3]
```diff
@@ -13,18 +13,19 @@ all: build test check-coverage lint  ## build, test, check coverage and lint
...
 
 install:  ## Build and install binaries in $GOBIN
-	go install -ldflags='$(GO_LDFLAGS)' .
+	go install -ldflags='$(GO_LDFLAGS)'  $(CMDS)
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply Makefile:28]: thanks,    done.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# A toolkit for working with protobuf and gRPC in Go
+
+[![ci](https://github.com/foxygoat/protog/actions/workflows/cicd.yaml/badge.svg?branch=master)](https://github.com/foxygoat/protog/actions/workflows/cicd.yaml?branch=master)
+[![Godoc](https://img.shields.io/badge/godoc-ref-blue)](https://pkg.go.dev/foxygo.at/protog)
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply README.md:4]: annoying, maybe give it some time and/or use this package and see if it resolves the issue?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+`make help`.
+
+## pb
+
+`pb` is a CLI tool for converting proto messages between different
+formats.
+
+Sample usage:
+
+	# create base-massage.pb binary encoded proto message file from input JSON
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply README.md:24]: sorry, no idea how/why. fixed. maybe cause i disabled prettier...

[the remark this answers is not in the document]
Comment (juliaogris) [reply README.md:24]: ok, ran prettier over it - that slightly updated whitespace.

File: cmd/pb/main.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	InFormat    string          `short:"I" help:"Input format (json, pb)" enum:"json,pb,j,p," default:""`
+	OutFormat   string          `short:"O" help:"Output format (j[son], p[b],t[xt])" enum:"json,pb,txt,j,p,t," default:""`
+	MessageType string          `arg:"" help:"Message type to be translated" required:""`
+	In          string          `arg:"" help:"Message value JSON encoded" optional:""`
+}
+
+func main() {
+	kctx := kong.Parse(&cli,
+		kong.Description(description),
+		kong.Vars{"version": fmt.Sprintf("%s (%s on %s)", version, commit, date)},
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:47]: they are set by goreleaser automatically.

[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:47]: removed LDFLAGS from makefile.

File: cmd/pb/main.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+		return err
+	}
+	b, err := marshal(message)
+	if err != nil {
+		return err
+	}
+	return cfg.writeOutput(b)
+}
+
+func (c *PBConfig) input() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:85]: yes, good idea, done.

File: cmd/pb/main.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	if strings.HasPrefix(c.In, "@") {
+		return os.ReadFile(c.In[1:])
+	}
+	return []byte(c.In), nil
+}
+
+func (c *PBConfig) writeOutput(b []byte) error {
+	if c.Out == "" {
+		if getFormat("", c.OutFormat) == "pb" && isTTY() {
+			return fmt.Errorf("cannot write binary to TTY")
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:98]: yes! thanks. much better, done.

File: cmd/pb/main.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+	ext := filepath.Ext(contentOrFile)
+	if len(ext) < 1 {
+		return ext
+	}
+	return canonicalFormat(ext[1:])
+}
+
+func canonicalFormat(format string) string {
+	switch format {
+	case "j", "json":
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:160]: fixed. yes.

File: cmd/pb/main.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+		return "json"
+	case "pb", "p":
+		return "pb"
+	case "txt", "t", "prototext", "prototxt":
+		return "txt"
+	}
+	return format
+}
+
+func lookupMessage(reg *registry.Files, message string) (protoreflect.MessageDescriptor, error) {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:170]: agreed, done.

File: cmd/pb/main.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+		return nil, fmt.Errorf("message not found: %s", message)
+	}
+	if len(result) > 1 {
+		return nil, fmt.Errorf("ambiguous message name: %s", message)
+	}
+	return result[0], nil
+}
+
+func lookupMessageInMD(md protoreflect.MessageDescriptor, message string) (mds []protoreflect.MessageDescriptor, exactMatch bool) {
+	if message == string(md.FullName()) {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:195]: yes, renamed vars a little too.

File: cmd/pb/main.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+	}
+	return mds, false
+}
+
+func registryMapper(kctx *kong.DecodeContext, target reflect.Value) error {
+	reg, ok := target.Interface().(*registry.Files)
+	if !ok {
+		panic("target is not a *registry.Files")
+	}
+	var fdsFile string
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:222]: yeah i know - I thought so too. 
i saw people at work do this, and only use `f` for `*os.File` as it usually a pretty temporary handle between name and content or func calls - so I thought I try this new pattern but I am happy to change again.

File: cmd/pb/testdata/pbtest.pb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/testdata/pbtest.pb:1]: good idea!

File: cmd/pb/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:222]: ok. after debates - agreed. changed.

File: cmd/pb/main.go [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+	}
+	files, err := registryFiles(fdsFile)
+	if err != nil {
+		return err
+	}
+	*reg = *files
+	return nil
+}
+
+func registryFiles(fdsFile string) (*registry.Files, error) {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go:234]: and this.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (juliaogris) [reply README.md:4]: i looked up what it takes to raise an issuse.... lots of blabla... (but fair enough).

[the remark this answers is not in the document]
Comment (camh-) [reply README.md:4]: yeah, leave it for now and we'll see what happens. Perhaps this PR should be adding "2022" to the license? Not sure what we're meant to do there.

File: cmd/pb/main.go [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+
+func getFormat(contentOrFile string, format string) string {
+	if format != "" {
+		return canonicalFormat(format)
+	}
+	if !strings.HasPrefix(contentOrFile, "@") {
+		return "json"
+	}
+	ext := filepath.Ext(contentOrFile)
+	if len(ext) < 1 {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply cmd/pb/main.go]: oups, so sorry! totally missed this. doesn't really make sense anyway imo. thank you rejigged.

**Review by camh- — APPROVED** [review]
:leaves: LGTM. I left one unresolved comment about the license. Feel free to ignore.

**Pull request merged**